### PR TITLE
Custom Search Domains

### PR DESCRIPTION
**Pull request by axier**
**What this PR does / why we need it**:
Adds Custom Search Domain on Linux and allows automatic dns updates on Windows DNS Server securely when custom vnet is used.
**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes # 
https://github.com/Azure/acs-engine/issues/1603

**Special notes for your reviewer**:

**If applicable**:
- [ ] documentation
- [ ] unit tests
- [x] tested backward compatibility (ie. deploy with previous version, upgrade with this branch)

**Release note**:

### CODE REVIEW

File: pkg/acsengine/engine.go [hunk 1]
```diff
@@ -454,6 +454,11 @@ func getParameters(cs *api.ContainerService, isClassicMode bool, generatorCode s
 	addValue(parametersMap, "fqdnEndpointSuffix", cloudSpecConfig.EndpointConfig.ResourceManagerVMDNSSuffix)
 	addValue(parametersMap, "targetEnvironment", GetCloudTargetEnv(location))
 	addValue(parametersMap, "linuxAdminUsername", properties.LinuxProfile.AdminUsername)
+	if properties.LinuxProfile.CustomSearchDomain != nil {
+		addValue(parametersMap, "sdName", properties.LinuxProfile.CustomSearchDomain.Name)
```
Comment (jackfrancis) [inline pkg/acsengine/engine.go]: Can we make the template param a little more descriptive? E.g., `customSearchDomainName` instead of `sdName`?

Comment (axier) [reply pkg/acsengine/engine.go]: Yes, I agree. And about the CLA signing I am waiting for legal.

**Comment by sylus** [comment]
Hi there I am successfully using the https://github.com/Azure/acs-engine/issues/1603 fix for Linux nodes so was curious if this P.R. attempts to address same thing? Only thing other issue is missing is support for Windows nodes.

File: examples/vnet/kubernetesvnet-customsearchdomain.json [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+        {
+          "name": "agentpri2",
+          "count": 2,
+          "vmSize": "Standard_D2_v2",
+          "vnetSubnetId": "/subscriptions/SUB_ID/resourceGroups/RG_NAME/providers/Microsoft.Network/virtualNetworks/VNET_NAME/subnets/SUBNET_NAME",
+          "availabilityProfile": "AvailabilitySet"
+        }
+      ],
+      "linuxProfile": {
+        "CustomSearchDomain": {
```
Comment (jackfrancis) [inline examples/vnet/kubernetesvnet-customsearchdomain.json]: this should be `customSearchDomain` (lowercase first letter)

File: docs/clusterdefinition.md [hunk 3]
```diff
@@ -454,6 +454,9 @@ A cluster can have 0 to 12 agent pool profiles. Agent Pool Profiles are used for
 |adminUsername|yes|describes the username to be used on all linux clusters|
 |ssh.publicKeys.keyData|yes|The public SSH key used for authenticating access to all Linux nodes in the cluster.  Here are instructions for [generating a public/private key pair](ssh.md#ssh-key-generation).|
 |secrets|no|specifies an array of key vaults to pull secrets from and what secrets to pull from each|
+|customsearchdomain.name|no|describes the search domain to be used on all linux clusters|
```
Comment (jackfrancis) [inline docs/clusterdefinition.md]: Let's use the exact case here, e.g., `customSearchDomain.name`

Thanks for including docs!

**Comment by jackfrancis** [comment]
@axier which part of this PR at present speaks to "allows automatic dns updates on Windows DNS Server securely when custom vnet is used"?

Thanks for this!

**Comment by axier** [comment]
Hi @jackfrancis, I'm using the realmUser and realmPassword to join the linux machine to active directory and sssd daemon to make dns record 
changes when windows dns server is in "onlysecure" mode. As it is, this is only useful when having a custom vnet.

**Comment by jackfrancis** [comment]
I don't think, strictly speaking, that custom VNET is required for this workflow. For example, the acs-engine-created VNET could be used after k8s cluster deployment to install the Windows DNS resources into. Right? Or will this implementation not work in that order?

**Comment by tesharp** [comment]
@jackfrancis From my experience with custom VNET and DNS I would think it is required in this case. If setting these parameters and not using a custom VNET it would not be able to resolve the other vm's, i.e. master cannot find agents etc. So the deployment of cluster would fail and you would never be able to create a Windows DNS resource.

@axier This is just a request from my side, but would it be possible to add support for realm password being a secret instead? Similar as how servicePrinicipal.secret is done perhaps. I would like to check in the template file in git, but not with passwords...

**Comment by axier** [comment]
Sorry for the late response, our company is interested in the integration of acs-engine in our existing infrastructure. 
@jackfrancis: In my opinion, I don't think it is interesting to deploy a dns server after the acs-engine deployment.
@jackfrancis: Functionally speaking, I am deploying a new cluster three times a day with our own compilation with this functionality and it works. But we only use it with Windows DNS Servers.
@tesharp: No problem, however I am a bit busy this week so I can't make the changes until next week.

**Comment by jackfrancis** [comment]
Let's aim to merge this tomorrow as-is. Follow-up improvements can come in as their own PRs. Let's not let this one get too stale. Thanks @axier!

And btw I'll help w/ the merge conflict tomorrow morning

**Comment by aminebizid** [comment]
What about this useful PR?

**Comment by axier** [comment]
@jackfrancis: I'm sorry for the late answer. I am going to resolve the conflicts and let me know if there is something else that needs to be done.

**Comment by jackfrancis** [comment]
@axier I've rebased against master, will run tests, and then merge if passed. Thanks!

**Comment by jackfrancis** [comment]
@CecileRobertMichon for code review
@axier I've moved some foo around a bit, could you kindly pull down these changes and perform some cluster tests w/ this new feature to confirm things still work?

Thanks!

File: parts/k8s/kubernetescustomscript.sh [hunk 4]
```diff
@@ -124,6 +126,10 @@ function installEtcd() {
     retrycmd_if_failure 10 1 5 sudo etcdctl member update $MEMBER ${ETCD_PEER_URL} || exit $ERR_ETCD_CONFIG_FAIL
 }
 
+function setupCustomSearchDomain() {
```
Comment (CecileRobertMichon) [inline parts/k8s/kubernetescustomscript.sh]: where is this function being called?

File: pkg/api/types.go [hunk 5]
```diff
@@ -759,6 +767,11 @@ func (l *LinuxProfile) HasSecrets() bool {
 	return len(l.Secrets) > 0
 }
 
+// HasSearchDomain returns true if the customer specified secrets to install
+func (l *LinuxProfile) HasSearchDomain() bool {
+	return l.CustomSearchDomain != nil
```
Comment (CecileRobertMichon) [inline pkg/api/types.go]: Maybe this should check the content (Name, RealmUser and Password) to make sure that they are not empty? other wise if my apimodel looks like: 

```

"linuxProfile": { 


 
+        "customSearchDomain": {}, 


 
+        "adminUsername": "azureuser", 


 
+        "ssh": { 


 
+          "publicKeys": [ 


 
+            { 


 
+              "keyData": "" 


 
+            } 


 
+          ] 


 
+        } 


 
+      }, 
```
acs-engine will try to apply the script. That or we shouls enforce validation to make sure that if CustomSearchDomain is specified it has a valid Name + User + Password combination.

File: pkg/api/vlabs/types.go [hunk 6]
```diff
@@ -581,6 +589,11 @@ func (a *AgentPoolProfile) SetSubnet(subnet string) {
 	a.subnet = subnet
 }
 
+// HasSearchDomain returns true if the customer specified secrets to install
+func (l *LinuxProfile) HasSearchDomain() bool {
+	return l.CustomSearchDomain != nil
```
Comment (CecileRobertMichon) [inline pkg/api/vlabs/types.go]: Same as above

**Pull request merged**